allennlp/data/dataloader.py

```diff
@@ -76,19 +76,24 @@ def __init__(
             worker_init_fn=worker_init_fn,
             multiprocessing_context=multiprocessing_context,
         )
-        self._batches_per_epoch = batches_per_epoch or super().__len__()
```

Comment:
I'm a bit confused about why this is different.  The only thing I can come up with is that there's something going on in the super class that makes it so that querying the length here doesn't work.  Can you give me some pointers on why this is fixing the issue?

Also, it'd be nice to add a test with lazy datasets that fails before this change and passes after the change, just to be sure we've actually captured the issue, and don't have a regression later.

Reply:
And @JohnGiorgi, can you confirm that this code solves your issue?

Reply:
@matt-gardner Looks like it does. Training now take the expected amount of time.

Reply:
@matt-gardner in short, the output of ```__len__``` method in case of ```_Lazy_Instances``` is not meaningful while ```self._batches_per_epoch``` has to be meaningful in order for the code to work.

Specifically, the problem with this line was that in the case of ```batches_per_epoch is None``` and ```_LazyInstances``` dataset, ```self._batches_per_epoch``` would be equal to ```1```. (```super().__len__()```  would return its overridden value of ```1```). 
And if ```self._batches_per_epoch``` is ```1``` then the code yields one batch per epoch which was the reason @JohnGiorgi had so short epochs.

Reply:
Ah, got it, thanks for the additional explanation.  To paraphrase, the problem is that `__iter__` will have the wrong behavior when you combine lazy datasets and no `_batches_per_epoch`.  `__len__` will still do the same thing, but `__iter__` only gives you one batch per epoch.